packages/react-reconciler/src/ReactFiberHooks.js

```diff
@@ -509,23 +505,21 @@ export function resetHooksAfterThrow(): void {
   ReactCurrentDispatcher.current = ContextOnlyDispatcher;
 
   if (didScheduleRenderPhaseUpdate) {
-    const current = (currentlyRenderingFiber: any).alternate;
-    if (current !== null) {
-      // There were render phase updates. These are only valid for this render
-      // pass, which we are now aborting. Remove the updates from the queues so
-      // they do not persist to the next render. We already did a single pass
-      // through the whole list of hooks, so we know that any pending updates
-      // must have been dispatched during the render phase. The ones that were
-      // dispatched before we started rendering were already transferred to the
-      // current hook's queue.
-      let hook: Hook | null = current.memoizedState;
-      while (hook !== null) {
-        const queue = hook.queue;
-        if (queue !== null) {
-          queue.pending = null;
-        }
-        hook = hook.next;
+    // There were render phase updates. These are only valid for this render
+    // phase, which we are now aborting. Remove the updates from the queues so
+    // they do not persist to the next render. Do not remove updates from hooks
+    // that weren't processed.
+    //
+    // Only reset the updates from the queue if it has a clone. If it does
+    // not have a clone, that means it wasn't processed, and the updates were
+    // scheduled before we entered the render phase.
```

Comment:
@sebmarkbage I think if we change the list to an array, we would just need some other mechanism to determine if it was processed or not. I added tests so we don't forget.